go/tools/gopackagesdriver/main.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"go/types"
+	"os"
+	"os/signal"
+	"strings"
+)
+
+type driverResponse struct {
+	NotHandled bool
+
+	// Sizes, if not nil, is the types.Sizes to use when type checking.
+	Sizes *types.StdSizes
+
+	// Roots is the set of package IDs that make up the root packages.
+	// We have to encode this separately because when we encode a single package
+	// we cannot know if it is one of the roots as that requires knowledge of the
+	// graph it is part of.
+	Roots []string `json:",omitempty"`
+
+	// Packages is the full set of packages in the graph.
+	// The packages are not connected into a graph.
+	// The Imports if populated will be stubs that only have their ID set.
+	// Imports will be connected and then type and syntax information added in a
+	// later pass (see refine).
+	Packages []*FlatPackage
+}
+
+var (
+	bazelBin          = getenvDefault("GOPACKAGESDRIVER_BAZEL", "bazel")
+	workspaceRoot     = os.Getenv("BUILD_WORKSPACE_DIRECTORY")
+	targets           = strings.Fields(os.Getenv("GOPACKAGESDRIVER_BAZEL_TARGETS"))
+	targetsQueryStr   = os.Getenv("GOPACKAGESDRIVER_BAZEL_QUERY")
+	targetsTagFilters = os.Getenv("GOPACKAGESDRIVER_BAZEL_TAG_FILTERS")
+)
+
+func getenvDefault(key, defaultValue string) string {
+	if v, ok := os.LookupEnv(key); ok {
+		return v
+	}
+	return defaultValue
+}
+
+func signalContext(parentCtx context.Context, signals ...os.Signal) (ctx context.Context, stop context.CancelFunc) {
+	ctx, cancel := context.WithCancel(parentCtx)
+	ch := make(chan os.Signal, 1)
+	go func() {
+		select {
+		case <-ch:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+	signal.Notify(ch, signals...)
+
+	return ctx, cancel
+}
+
+func run() error {
+	ctx, cancel := signalContext(context.Background(), os.Interrupt)
```

Comment:
Optional: use `signal.NotifyContext` instead. That was new in Go 1.16, so if you want to support lower versions for now, add a TODO to use it in the future when 1.16 is the minimum version.

Reply:
I started with it, then had to walk back to support older versions, will add a TODO for when 1.16 is the minimum version.